<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Katherine E. Dempsey">
<meta name="dcterms.date" content="2025-07-24">

<title>qPCR Fold Change Analysis Using RQdeltaCT – Katherine E. Dempsey</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-61eb52ac2311955f7c9ea7be0e571450.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Katherine E. Dempsey</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> <i class="bi bi-house-heart-fill" role="img">
</i> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../projects.html"> <i class="bi bi-clipboard-data" role="img">
</i> 
<span class="menu-text">Projects</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#purpose" id="toc-purpose" class="nav-link active" data-scroll-target="#purpose">Purpose</a>
  <ul class="collapse">
  <li><a href="#workflow" id="toc-workflow" class="nav-link" data-scroll-target="#workflow">Workflow</a></li>
  <li><a href="#installing-package" id="toc-installing-package" class="nav-link" data-scroll-target="#installing-package">0. Installing Package</a></li>
  <li><a href="#importing-tidying-data" id="toc-importing-tidying-data" class="nav-link" data-scroll-target="#importing-tidying-data">1. Importing &amp; Tidying Data</a>
  <ul class="collapse">
  <li><a href="#a.-quality-control" id="toc-a.-quality-control" class="nav-link" data-scroll-target="#a.-quality-control">1a. Quality Control</a></li>
  <li><a href="#b.-data-filtering" id="toc-b.-data-filtering" class="nav-link" data-scroll-target="#b.-data-filtering">1b. Data Filtering</a></li>
  </ul></li>
  <li><a href="#normalizing-data-to-reference-gene-δct" id="toc-normalizing-data-to-reference-gene-δct" class="nav-link" data-scroll-target="#normalizing-data-to-reference-gene-δct">2. Normalizing Data to Reference Gene (ΔCt)</a></li>
  <li><a href="#relative-quantification-δδct" id="toc-relative-quantification-δδct" class="nav-link" data-scroll-target="#relative-quantification-δδct">3. Relative Quantification (ΔΔCt)</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization">4. Visualization</a>
  <ul class="collapse">
  <li><a href="#a.-box-plot" id="toc-a.-box-plot" class="nav-link" data-scroll-target="#a.-box-plot">4a. Box Plot</a></li>
  <li><a href="#b.-bar-plot" id="toc-b.-bar-plot" class="nav-link" data-scroll-target="#b.-bar-plot">4b. Bar Plot</a></li>
  <li><a href="#c.-fold-change-plot" id="toc-c.-fold-change-plot" class="nav-link" data-scroll-target="#c.-fold-change-plot">4c. Fold Change Plot</a></li>
  </ul></li>
  <li><a href="#conclusion-3.55" id="toc-conclusion-3.55" class="nav-link" data-scroll-target="#conclusion-3.55">Conclusion: ⭐️ 3.5/5</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">qPCR Fold Change Analysis Using RQdeltaCT</h1>
<p class="subtitle lead">Intro to Reproducible Data Analysis Using R (CELL 6351) Final Project</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Katherine E. Dempsey </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 24, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="purpose" class="level1">
<h1>Purpose</h1>
<p>The goal of this project was to develop a well-documented script that streamlines the process of calculating and visualizing gene expression fold changes from raw qPCR datasets, entirely within R. For the purposes of this example, I will be showing qPCR data generated by a summer intern in my lab. The experiment is looking at the expression of polyamine metabolism genes in samples obtained from KPC mice treated with two different doses of pentamidine, which is an inhibitor of polyamine synthesis.</p>
<section id="workflow" class="level2">
<h2 class="anchored" data-anchor-id="workflow">Workflow</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://cran.r-project.org/web/packages/RQdeltaCT/vignettes/my-vignette.html#:~:text=The%20presented%20RQdeltaCT%20package%20includes,basic%20programming%20skills%20are%20required"><img src="RQdeltaCT_workflow.png" class="img-fluid figure-img"></a></p>
<figcaption><strong>Figure 1:</strong> Overview of workflow for <code>RQdeltaCT</code> package</figcaption>
</figure>
</div>
</section>
<section id="installing-package" class="level2">
<h2 class="anchored" data-anchor-id="installing-package">0. Installing Package</h2>
<p>You can view the technical details about this package on <a href="https://cran.r-project.org/web/packages/RQdeltaCT/index.html">CRAN</a> or <a href="https://github.com/Donadelnal/RQdeltaCT">GitHub</a>. Then follow the steps below to install and load the <code>RQdeltaCT</code> package. Follow the steps below to install from the CRAN repository.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"RQdeltaCT"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since you should already have already installed the <code>devtools</code> package for this class, you can also install <code>RQdeltaCT</code> directly from GitHub instead.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"Donadelnal/RQdeltaCT"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once you have installed <code>RQdeltaCT</code>, you can click <a href="https://cran.r-project.org/web/packages/RQdeltaCT/RQdeltaCT.pdf">here</a> to learn more about what functions are included in this package.</p>
</section>
<section id="importing-tidying-data" class="level2">
<h2 class="anchored" data-anchor-id="importing-tidying-data">1. Importing &amp; Tidying Data</h2>
<p>This package includes built-in <code>read_Ct_long()</code> and <code>read_Ct_wide()</code> functions for importing data, but these only recognize .csv and .txt files. However, most qPCR machines export results as an Excel spreadsheet (.xls, .xlsx). Below is a way to directly import specific sheets from Excel without having to create a separate file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The 'readxl' package allows you to open individual sheets from Excel without having to convert it into .csv or .txt format first</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"readxl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Your raw data file will probably look similar to the one below, which requires quite a bit of tidying.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="raw_qPCR_data_screenshot.png" class="img-fluid figure-img"></p>
<figcaption><strong>Figure 2:</strong> Example of raw qPCR data results</figcaption>
</figure>
</div>
<p>The first step is to import the sheet titled “Raw Data” from the Excel file using the <code>readxl</code> package. I also excluded the first 47 rows of irrelevant information pertaining to the qPCR instrument and protocol.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load necessary packages</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RQdeltaCT)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>raw_data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">"qPCR_data.xls"</span>, <span class="at">sheet =</span> <span class="st">"Results"</span>, <span class="at">skip =</span> <span class="dv">47</span>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(raw_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 121
Columns: 38
$ Well                     &lt;chr&gt; "151", "152", "153", "154", "155", "156", "15…
$ `Well Position`          &lt;chr&gt; "G7", "G8", "G9", "G10", "G11", "G12", "G13",…
$ Omit                     &lt;lgl&gt; FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FAL…
$ `Sample Name`            &lt;chr&gt; "c1", "c1", "c1", "c2", "c2", "c2", "c3", "c3…
$ `Target Name`            &lt;chr&gt; "18s", "18s", "18s", "18s", "18s", "18s", "18…
$ Task                     &lt;chr&gt; "UNKNOWN", "UNKNOWN", "UNKNOWN", "UNKNOWN", "…
$ Reporter                 &lt;chr&gt; "SYBR", "SYBR", "SYBR", "SYBR", "SYBR", "SYBR…
$ Quencher                 &lt;chr&gt; "None", "None", "None", "None", "None", "None…
$ Quantity                 &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `Quantity Mean`          &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `Quantity SD`            &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ RQ                       &lt;dbl&gt; NA, 7.1089945, 7.1089945, 6.8644419, 6.864441…
$ `RQ Min`                 &lt;dbl&gt; NA, 5.5418096, 5.5418096, 5.1459794, 5.145979…
$ `RQ Max`                 &lt;dbl&gt; NA, 9.119368, 9.119368, 9.156772, 9.156772, 9…
$ CT                       &lt;chr&gt; "Undetermined", "11.990301132202148", "11.887…
$ `Ct Mean`                &lt;dbl&gt; 11.93890, 11.93890, 11.93890, 11.89193, 11.89…
$ `Ct SD`                  &lt;dbl&gt; 0.07269488, 0.07269488, 0.07269488, 0.0913456…
$ `Delta Ct`               &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `Delta Ct Mean`          &lt;dbl&gt; NA, -14.87034, -14.87034, -14.81983, -14.8198…
$ `Delta Ct SD`            &lt;dbl&gt; NA, 0.18866599, 0.18866599, 0.25932723, 0.259…
$ `Delta Ct SE`            &lt;dbl&gt; NA, 0.11289690, 0.11289690, 0.14972264, 0.149…
$ `Delta Delta Ct`         &lt;dbl&gt; NA, -2.8296454, -2.8296454, -2.7791424, -2.77…
$ `Automatic Ct Threshold` &lt;lgl&gt; TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…
$ `Ct Threshold`           &lt;dbl&gt; 0.2658328, 0.2658328, 0.2658328, 0.2658328, 0…
$ `Automatic Baseline`     &lt;lgl&gt; TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…
$ `Baseline Start`         &lt;dbl&gt; 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, …
$ `Baseline End`           &lt;dbl&gt; 39, 7, 6, 7, 7, 7, 7, 8, 7, 8, 7, 8, 8, 8, 8,…
$ `Amp Status`             &lt;chr&gt; "No Amp", "Amp", "Amp", "Amp", "Amp", "Amp", …
$ Comments                 &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ `Cq Conf`                &lt;dbl&gt; 0.0000000, 0.9860583, 0.9873415, 0.9697938, 0…
$ CQCONF                   &lt;chr&gt; "Y", "N", "N", "N", "N", "N", "N", "N", "N", …
$ EXPFAIL                  &lt;chr&gt; "Y", "N", "N", "N", "N", "N", "N", "N", "N", …
$ NOAMP                    &lt;chr&gt; "N", "N", "N", "N", "N", "N", "N", "N", "N", …
$ Tm1                      &lt;dbl&gt; 84.77188, 83.41451, 83.39422, 83.39422, 83.50…
$ HIGHSD                   &lt;chr&gt; "N", "N", "N", "N", "N", "N", "N", "N", "N", …
$ Tm2                      &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ Tm3                      &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…
$ Tm4                      &lt;lgl&gt; NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, N…</code></pre>
</div>
</div>
<p>Using <code>dplyr</code>, I selected only the columns required for calculations and filtered out rows containing the “H2O” no template controls.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>select_data <span class="ot">&lt;-</span> raw_data <span class="sc">|&gt;</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="st">`</span><span class="at">Sample Name</span><span class="st">`</span>, <span class="st">`</span><span class="at">Target Name</span><span class="st">`</span>, CT) <span class="sc">|&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="st">`</span><span class="at">Sample Name</span><span class="st">`</span> <span class="sc">!=</span> <span class="st">"H2O"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(select_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 108
Columns: 3
$ `Sample Name` &lt;chr&gt; "c1", "c1", "c1", "c2", "c2", "c2", "c3", "c3", "c3", "5…
$ `Target Name` &lt;chr&gt; "18s", "18s", "18s", "18s", "18s", "18s", "18s", "18s", …
$ CT            &lt;chr&gt; "Undetermined", "11.990301132202148", "11.88749504089355…</code></pre>
</div>
</div>
<p>The <code>RQdeltaCT</code> package requires four columns of data: “Sample”, “Group”, “Gene”, and “Ct”. Therefore, it is necessary to re-name the column headings to ensure your data is recognized properly. I also had to add a “Group” column for grouping the biological replicates.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>clean_data <span class="ot">&lt;-</span> select_data <span class="sc">|&gt;</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">Sample =</span> <span class="st">`</span><span class="at">Sample Name</span><span class="st">`</span>, <span class="at">Gene =</span> <span class="st">`</span><span class="at">Target Name</span><span class="st">`</span>, <span class="at">Ct =</span> CT) <span class="sc">|&gt;</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Group =</span> <span class="fu">case_when</span>(</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">str_detect</span>(Sample, <span class="st">"c"</span>) <span class="sc">~</span> <span class="st">"Control"</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">str_detect</span>(Sample, <span class="st">"5"</span>) <span class="sc">~</span> <span class="st">"5uM"</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">str_detect</span>(Sample, <span class="st">"10"</span>) <span class="sc">~</span> <span class="st">"10uM"</span>))</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(clean_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 108
Columns: 4
$ Sample &lt;chr&gt; "c1", "c1", "c1", "c2", "c2", "c2", "c3", "c3", "c3", "5_1", "5…
$ Gene   &lt;chr&gt; "18s", "18s", "18s", "18s", "18s", "18s", "18s", "18s", "18s", …
$ Ct     &lt;chr&gt; "Undetermined", "11.990301132202148", "11.887495040893555", "11…
$ Group  &lt;chr&gt; "Control", "Control", "Control", "Control", "Control", "Control…</code></pre>
</div>
</div>
<p>You can also play around with sample data that is included with the package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(data.Ct)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(data.Ct)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   1288 obs. of  5 variables:
 $ Sample: chr  "AAA1" "AAA10" "AAA12" "AAA13" ...
 $ Gene  : chr  "ANGPT1" "ANGPT1" "ANGPT1" "ANGPT1" ...
 $ Ct    : chr  "32.563" "34.648" "35.059" "37.135" ...
 $ Group : chr  "AAA" "AAA" "AAA" "AAA" ...
 $ Flag  : chr  "OK" "OK" "OK" "OK" ...</code></pre>
</div>
</div>
<section id="a.-quality-control" class="level3">
<h3 class="anchored" data-anchor-id="a.-quality-control">1a. Quality Control</h3>
<p>First, let’s generate a few quick graphs to help us visualize the proportion of unreliable Ct values by flagging values that were undetermined or above the set threshold.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Evaluate the quality of reads for each sample</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>sample_quality <span class="ot">&lt;-</span> <span class="fu">control_Ct_barplot_sample</span>(<span class="at">data =</span> clean_data,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">flag.Ct =</span> <span class="st">"Undetermined"</span>, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">maxCt =</span> <span class="dv">35</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#68B2A0"</span>, <span class="st">"#FF7B89"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="RQdeltaCT_files/figure-html/Quality Check-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View tibble</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(sample_quality[[<span class="dv">2</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 4
  Sample Not.reliable Reliable Not.reliable.fraction
  &lt;fct&gt;         &lt;int&gt;    &lt;int&gt;                 &lt;dbl&gt;
1 c1                4        8                0.333 
2 10_2              3        9                0.25  
3 10_3              3        9                0.25  
4 5_1               3        9                0.25  
5 5_3               3        9                0.25  
6 c2                3        9                0.25  
7 5_2               2       10                0.167 
8 c3                2       10                0.167 
9 10_1              1       11                0.0833</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Evaluate the quality of reads for each gene</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>gene_quality <span class="ot">&lt;-</span> <span class="fu">control_Ct_barplot_gene</span>(<span class="at">data =</span> clean_data,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">flag.Ct =</span> <span class="st">"Undetermined"</span>,</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">maxCt =</span> <span class="dv">35</span>,</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#68B2A0"</span>, <span class="st">"#FF7B89"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="RQdeltaCT_files/figure-html/Quality Check-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View tibble</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gene_quality[[<span class="dv">2</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  Gene  Group   Not.reliable Reliable Not.reliable.fraction
  &lt;fct&gt; &lt;fct&gt;          &lt;int&gt;    &lt;int&gt;                 &lt;dbl&gt;
1 Arg1  5uM                8        1                 0.889
2 Arg1  Control            8        1                 0.889
3 Arg1  10uM               7        2                 0.778
4 18s   Control            1        8                 0.111
5 18s   10uM               0        9                 0    
6 Odc1  10uM               0        9                 0    </code></pre>
</div>
</div>
<p>Genes with consistently weak amplification can be an indicator of poor primer efficiency, meaning these primer pairs may need to be redesigned.</p>
</section>
<section id="b.-data-filtering" class="level3">
<h3 class="anchored" data-anchor-id="b.-data-filtering">1b. Data Filtering</h3>
<p>Next, let’s filter out the genes that had at least 50% low quality reads.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>low_quality_genes <span class="ot">&lt;-</span> <span class="fu">filter</span>(gene_quality[[<span class="dv">2</span>]], Not.reliable.fraction <span class="sc">&gt;</span> <span class="fl">0.5</span>)<span class="sc">$</span>Gene</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>low_quality_genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">as.vector</span>(low_quality_genes))</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>low_quality_genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Arg1"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Exclude identified genes from data set for analysis</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>quality_data <span class="ot">&lt;-</span> <span class="fu">filter_Ct</span>(<span class="at">data =</span> clean_data,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">remove.Gene =</span> low_quality_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally, we need to finish prepping data for analysis using the <code>make_Ct_ready()</code> function. Setting the argument <code>imput.by.mean.within.groups</code> to “TRUE” will collapse technical replicates by taking the average Ct values for each biological sample and remove any missing values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ready_data <span class="ot">&lt;-</span> <span class="fu">make_Ct_ready</span>(<span class="at">data =</span> quality_data,</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">imput.by.mean.within.groups =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The data contain 0 missing values that constitute 0 percent of the total data.
Missing values were imputed using means within compared groups.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>ready_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 5
# Groups:   Group [3]
  Group   Sample `18s`  Odc1  Sat1
  &lt;chr&gt;   &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 10uM    10_1    13.3  25.4  24.7
2 10uM    10_2    11.4  24.4  24.5
3 10uM    10_3    12.0  25.0  25.0
4 5uM     5_1     12.8  26.0  27.1
5 5uM     5_2     13.1  26.4  27.4
6 5uM     5_3     13.2  24.9  26.1
7 Control c1      11.9  26.8  26.1
8 Control c2      11.9  26.7  25.7
9 Control c3      12.3  27.0  26.3</code></pre>
</div>
</div>
</section>
</section>
<section id="normalizing-data-to-reference-gene-δct" class="level2">
<h2 class="anchored" data-anchor-id="normalizing-data-to-reference-gene-δct">2. Normalizing Data to Reference Gene (ΔCt)</h2>
<p>If you are calculating 2<sup>-dCt</sup>, then set the <code>transform</code> argument to “TRUE”. Otherwise, set it equal to “FALSE” if you are using the 2<sup>-ddCt</sup> method.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ΔCt method</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>deltaCt_results <span class="ot">&lt;-</span> <span class="fu">delta_Ct</span>(<span class="at">data =</span> ready_data,</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">normalise =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">ref =</span> <span class="st">"18s"</span>,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">transform =</span> <span class="cn">TRUE</span>)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>deltaCt_results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 4
# Groups:   Group [3]
  Group   Sample      Odc1      Sat1
  &lt;chr&gt;   &lt;chr&gt;      &lt;dbl&gt;     &lt;dbl&gt;
1 10uM    10_1   0.000237  0.000371 
2 10uM    10_2   0.000121  0.000117 
3 10uM    10_3   0.000123  0.000120 
4 5uM     5_1    0.000103  0.0000492
5 5uM     5_2    0.000101  0.0000514
6 5uM     5_3    0.000288  0.000124 
7 Control c1     0.0000334 0.0000528
8 Control c2     0.0000346 0.0000676
9 Control c3     0.0000371 0.0000602</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ΔΔCt method</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>deltaCt <span class="ot">&lt;-</span> <span class="fu">delta_Ct</span>(<span class="at">data =</span> ready_data,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">normalise =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">ref =</span> <span class="st">"18s"</span>,</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">transform =</span> <span class="cn">FALSE</span>)</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>deltaCt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 4
# Groups:   Group [3]
  Group   Sample  Odc1  Sat1
  &lt;chr&gt;   &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;
1 10uM    10_1    12.0  11.4
2 10uM    10_2    13.0  13.1
3 10uM    10_3    13.0  13.0
4 5uM     5_1     13.2  14.3
5 5uM     5_2     13.3  14.2
6 5uM     5_3     11.8  13.0
7 Control c1      14.9  14.2
8 Control c2      14.8  13.9
9 Control c3      14.7  14.0</code></pre>
</div>
</div>
<p>Another quality check can be implemented to observe the consistency between biological replicates and identify any outliers by generating a PCA plot using the <code>control_pca_sample</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>pca_check <span class="ot">&lt;-</span> <span class="fu">control_pca_sample</span>(deltaCt,</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">label.size =</span> <span class="dv">4</span>,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">plot.title =</span> <span class="st">"PCA Analysis"</span>,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">legend.title =</span> <span class="st">"Treatment"</span>,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>                                <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#FF7B89"</span>, <span class="st">"#A5678E"</span>, <span class="st">"#758EB7"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="RQdeltaCT_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="relative-quantification-δδct" class="level2">
<h2 class="anchored" data-anchor-id="relative-quantification-δδct">3. Relative Quantification (ΔΔCt)</h2>
<p>The <code>RQ_ddCt()</code> function calculates the relative fold change values along with statistical significance (either Student’s t test or Mann-Whitney U test is performed based on Shapiro-Wilk normality test). Therefore, the <code>RQ_ddCt()</code> function will need to be called for each group comparison you want to make, as specified by the <code>group.study</code> and <code>group.ref</code> arguments.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample size must be between 3 and 5000 for normality testing</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>ddCt_results_5uM <span class="ot">&lt;-</span> <span class="fu">RQ_ddCt</span>(deltaCt,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">group.study =</span> <span class="st">"5uM"</span>,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">group.ref =</span> <span class="st">"Control"</span>,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">do.tests =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(ddCt_results_5uM)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 2
Columns: 16
Groups: Gene [2]
$ Gene           &lt;chr&gt; "Odc1", "Sat1"
$ `5uM_mean`     &lt;dbl&gt; 12.75850, 13.84421
$ Control_mean   &lt;dbl&gt; 14.80289, 14.02729
$ `5uM_sd`       &lt;dbl&gt; 0.8645299, 0.7551560
$ Control_sd     &lt;dbl&gt; 0.07732601, 0.17846670
$ `5uM_norm_p`   &lt;dbl&gt; 0.03574696, 0.07928243
$ Control_norm_p &lt;dbl&gt; 0.6353363, 0.9231704
$ ddCt           &lt;dbl&gt; -2.0443874, -0.1830792
$ FCh            &lt;dbl&gt; 4.124981, 1.135304
$ log10FCh       &lt;dbl&gt; 0.61542193, 0.05511233
$ t_test_p       &lt;dbl&gt; 0.05370094, 0.71884188
$ t_test_stat    &lt;dbl&gt; -4.0795620, -0.4086593
$ MW_test_p      &lt;pvalue&gt; 0.04953461, 0.51269076
$ MW_test_stat   &lt;dbl&gt; -1.9639610, 0.6546537
$ t_test_p_adj   &lt;dbl&gt; 0.1074019, 0.7188419
$ MW_test_p_adj  &lt;dbl&gt; 0.09906923, 0.51269076</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>ddCt_results_10uM <span class="ot">&lt;-</span> <span class="fu">RQ_ddCt</span>(deltaCt,</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">group.study =</span> <span class="st">"10uM"</span>,</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">group.ref =</span> <span class="st">"Control"</span>,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">do.tests =</span> <span class="cn">TRUE</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(ddCt_results_10uM)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 2
Columns: 16
Groups: Gene [2]
$ Gene           &lt;chr&gt; "Odc1", "Sat1"
$ `10uM_mean`    &lt;dbl&gt; 12.68422, 12.49364
$ Control_mean   &lt;dbl&gt; 14.80289, 14.02729
$ `10uM_sd`      &lt;dbl&gt; 0.5574356, 0.9489090
$ Control_sd     &lt;dbl&gt; 0.07732601, 0.17846670
$ `10uM_norm_p`  &lt;dbl&gt; 0.03972178, 0.04339291
$ Control_norm_p &lt;dbl&gt; 0.6353363, 0.9231704
$ ddCt           &lt;dbl&gt; -2.118665, -1.533652
$ FCh            &lt;dbl&gt; 4.342920, 2.895179
$ log10FCh       &lt;dbl&gt; 0.6377818, 0.4616754
$ t_test_p       &lt;dbl&gt; 0.02064176, 0.10270610
$ t_test_stat    &lt;dbl&gt; -6.520629, -2.751153
$ MW_test_p      &lt;pvalue&gt; 0.04953461, 0.04953461
$ MW_test_stat   &lt;dbl&gt; -1.963961, -1.963961
$ t_test_p_adj   &lt;dbl&gt; 0.04128353, 0.10270610
$ MW_test_p_adj  &lt;dbl&gt; 0.04953461, 0.04953461</code></pre>
</div>
</div>
</section>
<section id="visualization" class="level2">
<h2 class="anchored" data-anchor-id="visualization">4. Visualization</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 'ggsignif' package allows you to add signficance labels to plots</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggsignif)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="a.-box-plot" class="level3">
<h3 class="anchored" data-anchor-id="a.-box-plot">4a. Box Plot</h3>
<p>Box plots are useful for visualizing the distribution of data, so here, I used the <code>results_boxplot</code> function to plot the distribution of raw Ct values from my <code>ready_data</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Can only graph ready_data or deltaCt</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>final_boxplot <span class="ot">&lt;-</span> <span class="fu">results_boxplot</span>(<span class="at">data =</span> ready_data,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">by.group =</span> <span class="cn">TRUE</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">signif.show =</span> <span class="cn">FALSE</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">faceting =</span> <span class="cn">TRUE</span>,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">facet.row =</span> <span class="dv">1</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">facet.col =</span> <span class="dv">3</span>,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#FF7B89"</span>, <span class="st">"#A5678E"</span>, <span class="st">"#758EB7"</span>),</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">plot.title =</span> <span class="st">"Distribution of Raw Ct Values"</span>,</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">y.axis.title =</span> <span class="st">"Ct"</span>,</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">legend.title =</span> <span class="st">"Treatment"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="RQdeltaCT_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="b.-bar-plot" class="level3">
<h3 class="anchored" data-anchor-id="b.-bar-plot">4b. Bar Plot</h3>
<p>Next, I used the <code>results_barplot</code> function to graph the transformed delta Ct values for each treatment group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Can only graph ready_data or deltaCt</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>final_barplot <span class="ot">&lt;-</span> <span class="fu">results_barplot</span>(deltaCt_results,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">signif.show =</span> <span class="cn">FALSE</span>,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">faceting =</span> <span class="cn">TRUE</span>,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">facet.row =</span> <span class="dv">1</span>,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">facet.col =</span> <span class="dv">2</span>,</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#FF7B89"</span>, <span class="st">"#A5678E"</span>, <span class="st">"#758EB7"</span>),</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">plot.title =</span> <span class="st">"Expression of Target Gene Normalized to Reference"</span>,</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">y.axis.title =</span> <span class="st">"2^-DeltaCt"</span>,</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">legend.title =</span> <span class="st">"Treatment"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="RQdeltaCT_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="c.-fold-change-plot" class="level3">
<h3 class="anchored" data-anchor-id="c.-fold-change-plot">4c. Fold Change Plot</h3>
<p>Finally, the <code>FCh_plot</code> function can be used to graph the relative fold changes. By plotting the values on a log scale, the expression levels are centered around 0, which represents the “Control” group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>FC_plot_5uM <span class="ot">&lt;-</span> <span class="fu">FCh_plot</span>(<span class="at">data =</span> ddCt_results_5uM,</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">use.p =</span> <span class="cn">TRUE</span>,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">mode =</span> <span class="st">"depends"</span>,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">p.threshold =</span> <span class="fl">0.05</span>,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">use.FCh =</span> <span class="cn">TRUE</span>,</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">signif.show =</span> <span class="cn">TRUE</span>,</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">signif.labels =</span> <span class="fu">c</span>(<span class="st">"*"</span>, <span class="st">"n.s."</span>),</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">signif.dist =</span> <span class="dv">0</span>,</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">signif.size =</span> <span class="dv">5</span>,</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#A5678E"</span>, <span class="st">"#758EB7"</span>),</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">plot.title =</span> <span class="st">"5 uM Dose Relative to Control"</span>,</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">plot.title.size =</span> <span class="dv">16</span>,</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">x.axis.title =</span> <span class="st">"Gene"</span>,</span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">axis.title.size =</span> <span class="dv">12</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="RQdeltaCT_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>FC_plot_10uM <span class="ot">&lt;-</span> <span class="fu">FCh_plot</span>(<span class="at">data =</span> ddCt_results_10uM,</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">use.p =</span> <span class="cn">TRUE</span>,</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">mode =</span> <span class="st">"depends"</span>,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">p.threshold =</span> <span class="fl">0.05</span>,</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">use.FCh =</span> <span class="cn">TRUE</span>,</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">signif.show =</span> <span class="cn">TRUE</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">signif.labels =</span> <span class="fu">c</span>(<span class="st">"*"</span>, <span class="st">" * "</span>),</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">signif.dist =</span> <span class="dv">0</span>,</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">signif.size =</span> <span class="dv">7</span>,</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#A5678E"</span>, <span class="st">"#758EB7"</span>),</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">plot.title =</span> <span class="st">"10 uM Dose Relative to Control"</span>,</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">plot.title.size =</span> <span class="dv">16</span>,</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">x.axis.title =</span> <span class="st">"Gene"</span>,</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">axis.title.size =</span> <span class="dv">12</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="RQdeltaCT_files/figure-html/unnamed-chunk-15-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Store calculations in data frame &amp; view selected p-values</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>FC_table_5uM <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(FC_plot_5uM[<span class="dv">2</span>])</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>FC_table_5uM <span class="sc">|&gt;</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Gene, FCh, log10FCh, test.for.comparison, p.used)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Gene      FCh   log10FCh test.for.comparison     p.used
1 Odc1 4.124981 0.61542193   Mann-Whitney.test 0.04953461
2 Sat1 1.135304 0.05511233    t.student's.test 0.71884188</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>FC_table_10uM <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(FC_plot_10uM[<span class="dv">2</span>])</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>FC_table_10uM <span class="sc">|&gt;</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Gene, FCh, log10FCh, test.for.comparison, p.used)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Gene      FCh  log10FCh test.for.comparison     p.used
1 Odc1 4.342920 0.6377818   Mann-Whitney.test 0.04953461
2 Sat1 2.895179 0.4616754   Mann-Whitney.test 0.04953461</code></pre>
</div>
</div>
<p>When you are evaluating many genes, log10(Fold Change) plots can be useful for quickly telling which ones have upregulated vs downregulated expression. Here is an example from another one of my datasets with more genes.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/clipboard-2793993996.png" class="img-fluid figure-img"></p>
<figcaption><strong>Figure 3:</strong> Fold change plot example</figcaption>
</figure>
</div>
</section>
</section>
<section id="conclusion-3.55" class="level2">
<h2 class="anchored" data-anchor-id="conclusion-3.55">Conclusion: ⭐️ 3.5/5</h2>
<p><strong>Pros</strong></p>
<ul>
<li><p>Detailed help sheet with example data and code</p></li>
<li><p>Quality control &amp; filtering abilities</p></li>
<li><p>Automatic normality testing and p-value calculations</p></li>
<li><p>Built-in graphing functions for generating quick plots</p></li>
</ul>
<p><strong>Cons</strong></p>
<ul>
<li><p>Sometimes requires extra data manipulation to achieve desired outcome</p></li>
<li><p>Less customization options for producing publication-quality plots</p></li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/katherined101\.github\.io\/Final_Project\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Copyright 2025, Katherine E. Dempsey</p>
</div>
  </div>
</footer>




</body></html>